src/grpc/stream_infer_handler.cc

```diff
@@ -359,13 +359,31 @@ ModelStreamInferHandler::Process(InferHandler::State* state, bool rpc_ok)
       response->mutable_infer_response()->Clear();
       // repopulate the id so that client knows which request failed.
       response->mutable_infer_response()->set_id(request.id());
-      state->step_ = Steps::WRITEREADY;
       if (!state->is_decoupled_) {
+        state->step_ = Steps::WRITEREADY;
         state->context_->WriteResponseIfReady(state);
       } else {
-        state->response_queue_->MarkNextResponseComplete();
-        state->complete_ = true;
-        state->context_->PutTaskBackToQueue(state);
+        InferHandler::State* writing_state = nullptr;
+        std::lock_guard<std::recursive_mutex> lk1(state->context_->mu_);
+        {
+          std::lock_guard<std::recursive_mutex> lk2(state->step_mtx_);
+          bool has_prev_ready_response =
+              state->response_queue_->HasReadyResponse();
+          state->response_queue_->MarkNextResponseComplete();
+          if (!has_prev_ready_response) {
+            state->context_->ready_to_write_states_.push(state);
+          }
+          if (!state->context_->ongoing_write_ &&
+              !state->context_->ready_to_write_states_.empty()) {
+            state->context_->ongoing_write_ = true;
```

Comment:
Can you add comments here describing why we are circulating through the ready_to_write_states_ when there is no ongoing_write_? 
Describe how the state would be picked up later once the ongoing_write_ is completed.

Reply:
Also mention that ongoing_write would be reset once the completion queue returns a written state object.

Reply:
Sure, [Add comments on ready to write lifecycle](https://github.com/triton-inference-server/server/pull/7404/commits/edfaa2ee71eec8bed692feeefa79c56e8d81578a)